### ISSUE TITLE
Should the encoded initialisation vector be moved to the start?

### ISSUE THREAD

**Issue by Pharap** [gioblu/Cape#16]
I've been thinking.

The iv has to be recovered before the data can be decrypted, but at the moment it's at the end of the data - so the last byte has to be read before the first byte, which is inefficient.

If the iv was moved to the start of the encrypted data, then decrypting might be more efficient because the whole operation would then be one traversal through the data instead of skipping to the end and then backtracking.

(It would also allow more efficient implementations in languages that use (one-way) streams or iterators.)

**Comment by gioblu** [comment]
Ciao @Pharap thank you for your support, it should be considered how much we can save doing this, because it is breaking compatibility, although I have in any case to release 3.0 soon!

**Comment by gioblu** [comment]
@Pharap I am not sure, because the way the iv is encrypted is different from the way the plaintext is encrypted!

**Comment by Pharap** [comment]
@gioblu Firstly, compatibility is already broken (hence the next version is 3.0) so a few changes won't hurt.

Secondly, they _are_ encrypted the same way.

The current system:
```
destination[length] = iv ^ length ^ _key[(length ^ srk) % _key_length];
for(uint16_t i = 0; i < length; ++i)
destination[i] = source[i] ^ iv ^ i ^ _key[(srk ^ i) % _key_length];
```

Is equivalent to:
```
for(uint16_t i = 0; i < length + 1; ++i)
    if(i != length)
        destination[i] = source[i] ^ iv ^ i ^ _key[(srk ^ i) % _key_length];
    else
        destination[i] = 0 ^ iv ^ i ^ _key[(srk ^ i) % _key_length];
```

Or if `source[length] == 0` then:
```
for(uint16_t i = 0; i < length + 1; ++i)
    destination[i] = source[i] ^ iv ^ i ^ _key[(srk ^ i) % _key_length];
```

**Comment by gioblu** [comment]
Ciao @Pharap thank you for dedicating effort to Cape 👍 .
I agree with you, if more efficient, considering the compatibility has been broken already it would have sense to include this change!

**Comment by Pharap** [comment]
I did some tests and this approach ended up being slower for reasons I'm still unsure of.
It's just under 30 microseconds slower, but uses 6 less bytes of progmem.

(Times are in microseconds)
|  Method  | Best Encrypt Time | Worst Encrypt Time | Best Decrypt time | Worst Decrypt Time |
| --------- | ------------- | -------------- | -------------- | -------------- |
| iv at end  | 188 | 196 | 172 | 180 |
| iv at start | 184 | 196 | 200 | 204 |

**Comment by gioblu** [comment]
Ciao @Pharap I think it maybe better to keep it as it is now, not necessarily breaking compatibility, considering your test results. What do you think?

**Comment by Pharap** [comment]
I've been doing a few more tests.

Firstly I found out why the time was different - there was an off by one error in my decrypt implementation that meant the loop was doing an extra iteration that was going past the end of the array (I forgot to use `length - 1` instead of `length` when I tweaked the loop body).

So the best decrypt time for 'iv at start' is actually 184 and the worst decrypt time is 192.
It's still slower, but not quite as drastic, it's more like 10 microseconds rather than 20.

That particular version was using this:

https://gist.github.com/Pharap/1cef50c1432ce28fc2cbc68fc770e89e

Since then I've tried a slightly altered version:

https://gist.github.com/Pharap/5cbf7bc3ceafc241caefc56219c06a6b

Which saves 26 bytes and runs at similar speeds.
The revised figures are:

Original
Size Of Sketch: 4716
Best Encrypt Time: 188
Worst Encrypt Time: 196
Best Decrypt Time: 172
Worst Decrypt Time: 180

Rev1
Size Of Sketch: 4710
Best Encrypt Time: 184
Worst Encrypt Time: 196
Best Decrypt Time: 184
Worst Decrypt Time: 192

Rev2
Size Of Sketch: 4690
Best Encrypt Time: 188
Worst Encrypt Time: 196
Best Decrypt Time: 176
Worst Decrypt Time: 184

I've only tested on a Leonardo-based board though (the CPU is an ATmega 32u4).

Here's the code for the test I used so you can check with some different boards and draw your own conclusions:

https://gist.github.com/Pharap/b7e2a6bba1d183dafd178f9eba939153

The other benefits are:
* It's easier to implement in certain languages. (In C++ using iterators, in C# using enumerables, in Haskell/Lisp/ML using cons lists)
* It's better for file streaming implementations because there's no need to backtrack and the whole file doesn't have to be read at once
* It's possibly better for CPUs that have caches because the lack of backtracking means the cache won't have to be filled with the end of the array and then disposed of after reading just one byte

---

I'll leave you to make up your own mind.

If you're not satisfied that the figures are beneficial enough to consider changing the system or you are concerned that it would somehow be making the system less secure, then by all means close the issue and forget about it.

**Comment by gioblu** [comment]
Ciao @Pharap could you please explain me in more detail what you mean with the point 2?

**Comment by Pharap** [comment]
Files are typically accessed sequentially from beginning to end.
By placing the encoded IV at the end, the code would have to seek to the end to read it and then seek back to the start to finish doing the decryption.

To give an example using C#'s [FileStream class](https://msdn.microsoft.com/en-us/library/system.io.filestream(v=vs.110).aspx) the code would have to call `Seek(length, SeekOrigin.Current)` to get to the end, `ReadByte()` to read the encrypted IV and then `Seek(-length, SeekOrigin.Current)` to get back to the start of the data.
e.g.
```
stream.Seek(length, SeekOrigin.Current);
byte encodedIV = stream.ReadByte();
stream.Seek(-length, SeekOrigin.Current);
// Continue decrypting
```

Whereas putting it at the start would eliminate both these seek operations.
Seeking is cheap for flash-based storage like SSDs or on-board flash memory, but for hard drives it's a an issue because it means physically moving the read/write arm which can be slow even if the disk itself is fast.

I can't find many figures, but [one SO answer](https://stackoverflow.com/a/38029659/1377706) claims that for a 7200rpm HDD it would take around 8ms per seek, and for an SSD around 0.3ms per seek.
[Someone over at perlmonks](http://www.perlmonks.org/?node_id=923355) estimates 5ms per seek on a high-end HDD.
It's a complicated thing to guess because there's usually layers of buffering involved among other things.

Admittedly this probably won't be an issue for small amounts of data where the entire file will fit in memory, it would probably only be an issue for large files and even then it's only a small speed bump, so it may be that such cases are 'beyond the scope' of Cape.
I think it's worth consideration though, some people still use old hardware and it would be nice to tell them that they have been considered.

**Comment by Pharap** [comment]
When this has been resolved (whether it's accepted or not) I have two Cape ports ready to go.
One for C# and one for Haskell, each in a single code file and each set up to work as a library.
(I'd rather wait until the 3.0 release before publishing them so I can say "a port of Cape 3.0" instead of "based on Cape at this point in history".)

**Comment by gioblu** [comment]
Ciao @Pharap, that is amazing :). I was making some tests to be sure. Let's release Cape as it is. We can always release a 4.0 in future if this change will be required. I will close here and release tonight 3.0.

**Comment by Pharap** [comment]
Fair enough.
If there are definitely going to be no more changes to the code then I'll get my ports ready.

The only thing left to resolve is how the licencing is handled.
Ports tend to be a bit awkward when it comes to licencing since they are technically derivative works but often look nothing like the original.

Should I put the original licence notice at the top and then state modifications underneath?
Or should I put two licence headers, one for 'Cape' and one for 'name of cape port' (both being the same Apache 2.0 licence, the first being "Cape Copyright (c) 2012-2018, , Giovanni Blu Mitolo All rights reserved." and the second being "'name of cape port' Copyright (c) 2018, Pharap")?

I looked at the other two ports to see what they did but one doesn't include any licence and the other has the original licence header but neglects to state the changes, so they weren't much help.

**Comment by Pharap** [comment]
I decided to go for two licence headers.
It's longer, but I think it makes it clearer that there are two entities involved (the original and the port) and both use the same licence.

I've produced two ports:
* [CapeDotNet](https://github.com/Pharap/CapeDotNet/blob/master/Cape.cs) in C#.
* [CapeHaskell](https://github.com/Pharap/CapeHaskell/blob/master/Cape.hs) in Haskell.

I deviated slightly from the original:
* CapeDotNet doesn't have a `SetKey` function. I could add one if need be, but C# is unlikely to be found on low-memory devices so it wouldn't have much benefit.
* CapeDotNet has some `IEnumerable` generator functions.
* CapeHaskell doesn't have a `setKey` function because objects in Haskell aren't mutable, a `setKey` function would effectively be returning a new object anyway, so using the constructor seems as easy.
* CapeHaskell currently doesn't have a means of retrieving or modifying the salt, so I might have to address that later.
